### PR TITLE
Wrap partitioners within MatrixFree

### PR DESCRIPTION
**Pull request by peterrum**
Depends on #11074.

### CODE REVIEW

File: source/matrix_free/vector_data_exchange.cc [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+#include <deal.II/matrix_free/vector_data_exchange.h>
+
+#ifdef DEAL_II_WITH_64BIT_INDICES
+#  include <deal.II/base/mpi_consensus_algorithms.templates.h>
+#endif
+
+#include <map>
+#include <vector>
+
+#define DO_COMPRESS true
```
Comment (kronbichler) [inline source/matrix_free/vector_data_exchange.cc:32]: Is this still needed?

File: source/matrix_free/vector_data_exchange.cc [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/mpi.templates.h>
+#include <deal.II/base/mpi_compute_index_owner_internal.h>
+#include <deal.II/base/mpi_consensus_algorithms.h>
+#include <deal.II/base/timer.h>
+
+#include <deal.II/matrix_free/vector_data_exchange.h>
+
+#ifdef DEAL_II_WITH_64BIT_INDICES
+#  include <deal.II/base/mpi_consensus_algorithms.templates.h>
```
Comment (kronbichler) [inline source/matrix_free/vector_data_exchange.cc:26]: I guess you need this in a later PR?

**Review by kronbichler — APPROVED** [review]
Looks good to me.

File: source/matrix_free/vector_data_exchange.cc (hunk 1 above)
Comment (peterrum) [reply source/matrix_free/vector_data_exchange.cc]: I have remove this.

File: source/matrix_free/vector_data_exchange.cc (hunk 2 above)
Comment (peterrum) [reply source/matrix_free/vector_data_exchange.cc:26]: Yes, it is need by a follow-up PR.

File: include/deal.II/matrix_free/vector_data_exchange.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+
+      /**
+       * Class that simply delegates the task to a Utilities::MPI::Partitioner.
+       */
+      class PartitionerWrapper : public Base
+      {
+      public:
+        PartitionerWrapper(
+          const std::shared_ptr<const Utilities::MPI::Partitioner> partitioner);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/vector_data_exchange.h:146]: To make `clang-tidy` happy, you should pass this by const reference.

Comment (peterrum) [reply include/deal.II/matrix_free/vector_data_exchange.h]: I have made the modification!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**